src/ir/possible-contents.cpp

```diff
@@ -0,0 +1,1610 @@
+    }
+    return true;
+  }
+
+  bool isRelevant(Signature sig) {
+    return isRelevant(sig.params) || isRelevant(sig.results);
+  }
+
+  bool isRelevant(Expression* curr) { return curr && isRelevant(curr->type); }
+
+  template<typename T> bool isRelevant(const T& vec) {
+    return true;
+    for (auto* expr : vec) {
+      if (isRelevant(expr->type)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  // Each visit*() call is responsible for connecting the children of a node to
+  // that node. Responsibility for connecting the node's output to anywhere
+  // else (another expression or the function itself, if we are at the top
+  // level) is the responsibility of the outside.
+
+  void visitBlock(Block* curr) {
+    if (curr->list.empty()) {
+      return;
+    }
+
+    // Values sent to breaks to this block must be received here.
+    handleBreakTarget(curr);
+
+    // The final item in the block can flow a value to here as well.
+    receiveChildValue(curr->list.back(), curr);
+  }
+  void visitIf(If* curr) {
+    // Each arm may flow out a value.
+    receiveChildValue(curr->ifTrue, curr);
+    receiveChildValue(curr->ifFalse, curr);
+  }
+  void visitLoop(Loop* curr) { receiveChildValue(curr->body, curr); }
+  void visitBreak(Break* curr) {
+    // Connect the value (if present) to the break target.
+    handleBreakValue(curr);
+
+    // The value may also flow through in a br_if (the type will indicate that,
+    // which receiveChildValue will notice).
+    receiveChildValue(curr->value, curr);
+  }
+  void visitSwitch(Switch* curr) { handleBreakValue(curr); }
+  void visitLoad(Load* curr) {
+    // We could infer the exact type here, but as no subtyping is possible, it
+    // would have no benefit, so just add a generic root (which will be a Many).
+    addRoot(curr);
+  }
+  void visitStore(Store* curr) {}
+  void visitAtomicRMW(AtomicRMW* curr) { addRoot(curr); }
+  void visitAtomicCmpxchg(AtomicCmpxchg* curr) { addRoot(curr); }
+  void visitAtomicWait(AtomicWait* curr) { addRoot(curr); }
+  void visitAtomicNotify(AtomicNotify* curr) { addRoot(curr); }
+  void visitAtomicFence(AtomicFence* curr) {}
+  void visitSIMDExtract(SIMDExtract* curr) { addRoot(curr); }
+  void visitSIMDReplace(SIMDReplace* curr) { addRoot(curr); }
+  void visitSIMDShuffle(SIMDShuffle* curr) { addRoot(curr); }
+  void visitSIMDTernary(SIMDTernary* curr) { addRoot(curr); }
+  void visitSIMDShift(SIMDShift* curr) { addRoot(curr); }
+  void visitSIMDLoad(SIMDLoad* curr) { addRoot(curr); }
+  void visitSIMDLoadStoreLane(SIMDLoadStoreLane* curr) { addRoot(curr); }
+  void visitMemoryInit(MemoryInit* curr) {}
+  void visitDataDrop(DataDrop* curr) {}
+  void visitMemoryCopy(MemoryCopy* curr) {}
+  void visitMemoryFill(MemoryFill* curr) {}
+  void visitConst(Const* curr) {
+    addRoot(curr, PossibleContents::literal(curr->value));
+  }
+  void visitUnary(Unary* curr) { addRoot(curr); }
+  void visitBinary(Binary* curr) { addRoot(curr); }
```

Comment:
Can't we infer `ExactType` if the type of these don't have subtypes, like `i32`? The same for most other `addRoot`s as well. Maybe I'm mistaken about some assumptions.

Reply:
Hmm, it probably is worth an extra comment here - since this "use Many" principle is used in all these `addRoot` places. I'll add one here, which will hopefully address the previous comment too.

Reply:
I ended up writing a longer comment at the top of the header, and linking to it. The old comment was actually refactored together with some code into the cpp, so there was no longer a good comment in the header, which is the right place for it... fixed.